src/tmpfiles/tmpfiles.c

```diff
@@ -195,17 +208,54 @@ static int specifier_machine_id_safe(char specifier, void *data, void *userdata,
         return r;
 }
 
+static int specifier_directory(char specifier, void *data, void *userdata, char **ret) {
+        static const struct mapping {
+                const char *variable;
+                const char *system_value;
+                const char *suffix;
+        } mapping[] = {
+                [DIRECTORY_RUNTIME] = { "XDG_RUNTIME_DIR", "/run",       NULL   },
+                [DIRECTORY_STATE] =   { "XDG_CONFIG_HOME", "/var/lib",   NULL   },
+                [DIRECTORY_CACHE] =   { "XDG_CACHE_HOME",  "/var/cache", NULL   },
+                [DIRECTORY_LOGS] =    { "XDG_CONFIG_HOME", "/var/log",   "/log" },
+        };
+
+        unsigned index;
+        const char *s;
+        char *t;
+
+        index = PTR_TO_UINT(data);
+        assert(index < ELEMENTSOF(mapping));
+        assert(mapping[index].variable);
+
+        if (arg_user) {
+                s = getenv(mapping[index].variable);
```

Comment:
this doesn't look right... we need a fallback here for XDG_*_HOME to use the default paths... after all XDG_*_HOME are unlikely to be set by default

i.e. use sd_path_home() here really, which knows the rights fallbacks. See how manager.c does that too

Reply:
Reworked.